### PR TITLE
Applications: Create a display properties manager

### PR DESCRIPTION
**Pull request by Quaker762**
An interactive application to modify the current display settings, such as
the current wallpaper as well as the screen resolution. Currently we're
adding the resolutions ourselves, because there's currently no way to
detect was resolutions the current display adapter supports (or at least
I can't see one... Maybe VBE does and I'm stupid). It even comes with
a very nice template'd `ItemList` that can support a vector of any type,
which makes life much simpler.

This currently doesn't support modifying the actual theme of the system (i.e window border colours), but I can quickly whip that up later if everybody wants it.

### CODE REVIEW

File: Applications/DisplayProperties/DisplayProperties.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    int width;
+    int height;
+};
+
+class DisplayPropertiesWidget final
+{
+public:
+    enum class ButtonOperations
+    {
+        OK,
```
Comment (awesomekling) [inline Applications/DisplayProperties/DisplayProperties.h]: Enums should be InterCaps, not ALLCAPS.
https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md

File: Applications/DisplayProperties/DisplayProperties.cpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        apply_settings();
+    };
+
+    auto* cancel_button = new GButton(bottom_widget);
+    cancel_button->set_text("Cancel");
+    cancel_button->set_size_policy(Orientation::Vertical, SizePolicy::Fixed);
+    cancel_button->set_size_policy(Orientation::Horizontal, SizePolicy::Fixed);
+    cancel_button->set_preferred_size(60, 22);
+    cancel_button->on_click = [this](GButton&) {
+        GApplication::the().quit(EXIT_SUCCESS);
```
Comment (awesomekling) [inline Applications/DisplayProperties/DisplayProperties.cpp]: You can omit the `EXIT_SUCCESS` here.
```suggestion
        GApplication::the().quit();
```

File: Servers/WindowServer/WSAPITypes.h [hunk 3]
```diff
@@ -260,6 +263,21 @@ struct WSAPI_ClientMessage {
             bool minimized;
             WSAPI_Point position;
         } wm;
+        struct {
+            WSAPI_Size resolution;
+            u32 active_win_col;
```
[the remark this answers is not in the document]
Comment (awesomekling) [reply Servers/WindowServer/WSAPITypes.h]: Also, this seems unrelated to the patch at hand.

File: Applications/DisplayProperties/DisplayProperties.cpp [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+void DisplayPropertiesWidget::create_resolution_list()
+{
+    m_resolutions.append({640, 480});
+    m_resolutions.append({800, 600});
+    m_resolutions.append({1024, 768});
+    m_resolutions.append({1280, 1024});
+    m_resolutions.append({1366, 768});
+
+    m_selected_res = m_resolutions.at(0);
```
Comment (bugaevc) [inline Applications/DisplayProperties/DisplayProperties.cpp]: This is problematic. The app starts up with 640x480 selected *on the second tab*, but say you just want to change the wallpaper so you never even switch to the second tab. You pick the wallpaper you like, press _Apply_ and suddenly your whole QEMU window shrinks.

**Comment by bugaevc** [comment]
I was going to do a (somewhat more ambitious) Settings app sometime; this is a good starting point. Do you think it would make sense to name it Settings now (or I could rename it later)?

The idea was to have one app that would let you tweak all sorts of system settings, for WindowServer, LibGUI, Launcher, network, users (`/etc/passwd`), perhaps even file systems (`/etc/fstab`). Somewhat like the Control Panel on Windows.

File: Applications/DisplayProperties/DisplayProperties.cpp (hunk 4 above)
Comment (Quaker762) [reply Applications/DisplayProperties/DisplayProperties.cpp]: > This is problematic. The app starts up with 640x480 selected on the second tab, but say you just want to change the wallpaper so you never even switch to the second tab. 

Whoops, I forgot I had that there. I couldn't really come up with a good method of autodetection, though maybe have `m_resolutions.at(0)` be the currently selected resolution?

**Comment by Quaker762** [comment]
> I was going to do a (somewhat more ambitious) Settings app sometime; this is a good starting point. Do you think it would make sense to name it Settings now (or I could rename it later)?
> 
> The idea was to have one app that would let you tweak all sorts of system settings, for WindowServer, LibGUI, Launcher, network, users (`/etc/passwd`), perhaps even file systems (`/etc/fstab`). Somewhat like the Control Panel on Windows.

Yeah sure, that sounds pretty good, though would it make more sense to have each settings as individual modules? That is, if we're going with the Windows Control Panel route, Control Panel itself is a process that just launches other processes etc

File: Applications/DisplayProperties/DisplayProperties.cpp [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+    while(iterator.has_next())
+        m_wallpapers.append(iterator.next_path());
+}
+
+void DisplayPropertiesWidget::create_frame()
+{
+    auto* tabWidget = new GTabWidget(m_root_widget);
+
+    // First, let's create the "Background" page in the tabbed journal
```
[the remark this answers is not in the document]
Comment (Quaker762) [reply Applications/DisplayProperties/DisplayProperties.cpp]: Sorry I think I was getting this confused with the terminology used the one used in KDE or GTK

File: Servers/WindowServer/WSAPITypes.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (Quaker762) [reply Servers/WindowServer/WSAPITypes.h]: I'm going to remove this for now then and re-add it when I have time to work on the theme stuff :)

**Comment by Quaker762** [comment]
> Do you think it would make sense to name it Settings now (or I could rename it later)?

I've kept this as `DisplayProperties` for now, but when I've got time I'll change everything to `Control Panel`, `System Settings` or something to that effect :P

File: Applications/DisplayProperties/DisplayProperties.cpp (hunk 4 above)
Comment (awesomekling) [reply Applications/DisplayProperties/DisplayProperties.cpp]: How about we only *apply* the settings in the visible tab?

Comment (Quaker762) [reply Applications/DisplayProperties/DisplayProperties.cpp]: > How about we only _apply_ the settings in the visible tab?

I like that idea. I've modified `GTabWidget` accordingly to allow for acquiring the currently active tab (as an index).

File: Libraries/LibGUI/GTabWidget.h [hunk 6]
```diff
@@ -15,6 +15,7 @@ class GTabWidget : public GWidget {
 
     TabPosition tab_position() const { return m_tab_position; }
     void set_tab_position(TabPosition);
+    int get_active_tab();
```
Comment (awesomekling) [inline Libraries/LibGUI/GTabWidget.h]: Please add this API in a separate commit, so it's isolated from the other changes.
Also, it should return be `const` and maybe have a slightly more descriptive name like:

```cpp
int active_tab_index() const;
```

File: Servers/WindowServer/WSAPITypes.h [hunk 7]
```diff
@@ -224,6 +225,8 @@ struct WSAPI_ClientMessage {
         Greeting,
         SetWallpaper,
         GetWallpaper,
+        SetResolution,
+        GetResolution,
```
Comment (awesomekling) [inline Servers/WindowServer/WSAPITypes.h]: I don't see this used anywhere.

File: Applications/DisplayProperties/ItemListModel.h [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        return {};
+    }
+
+    virtual void update() override
+    {
+        did_update();
+    }
+
+private:
+    explicit ItemListModel(AK::Vector<T>& data)
```
Comment (awesomekling) [inline Applications/DisplayProperties/ItemListModel.h]: No need for `AK::` here.

File: Applications/DisplayProperties/DisplayProperties.h [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    void create_root_widget();
+
+private:
+    String m_wallpaper_path;
+    RefPtr<CConfigFile> m_wm_config;
+    GWidget* m_root_widget { nullptr };
+    Vector<Size> m_resolutions;
+    Vector<String> m_wallpapers;
+
+    Size m_selected_res;
```
Comment (awesomekling) [inline Applications/DisplayProperties/DisplayProperties.h]: Please spell things out, so `m_selected_resolution`.

File: Applications/DisplayProperties/DisplayProperties.cpp [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+{
+    CDirIterator iterator("/res/wallpapers/", CDirIterator::Flags::SkipDots);
+
+    while (iterator.has_next())
+        m_wallpapers.append(iterator.next_path());
+}
+
+void DisplayPropertiesWidget::create_frame()
+{
+    auto* tabWidget = new GTabWidget(m_root_widget);
```
Comment (awesomekling) [inline Applications/DisplayProperties/DisplayProperties.cpp]: `tab_widget`

File: Applications/DisplayProperties/DisplayProperties.cpp [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    m_resolutions.append({ 1600, 900 });
+    m_resolutions.append({ 1920, 1080 });
+    m_resolutions.append({ 2560, 1080 });
+
+    m_selected_res = m_resolutions.at(0);
+}
+
+void DisplayPropertiesWidget::create_root_widget()
+{
+    auto* keeper = new GWidget;
```
Comment (awesomekling) [inline Applications/DisplayProperties/DisplayProperties.cpp]: No need for a temporary here, really.
```suggestion
    m_root_widget = new GWidget;
```

File: Applications/DisplayProperties/DisplayProperties.cpp [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+    keeper->set_layout(make<GBoxLayout>(Orientation::Vertical));
+    keeper->set_fill_with_background_color(true);
+    keeper->set_background_color(Color::WarmGray);
```
Comment (awesomekling) [inline Applications/DisplayProperties/DisplayProperties.cpp]: This is already the default background color.

File: Applications/DisplayProperties/DisplayProperties.cpp [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+    cancel_button->set_text("Cancel");
+    cancel_button->set_size_policy(Orientation::Vertical, SizePolicy::Fixed);
+    cancel_button->set_size_policy(Orientation::Horizontal, SizePolicy::Fixed);
+    cancel_button->set_preferred_size(60, 22);
+    cancel_button->on_click = [this](GButton&) {
+        GApplication::the().quit();
+    };
+}
+
+void DisplayPropertiesWidget::send_settings_to_window_server(int tabIndex)
```
Comment (awesomekling) [inline Applications/DisplayProperties/DisplayProperties.cpp]: `tab_index`

**Comment by awesomekling** [comment]
All right, let's take it!

**Pull request merged**